daemons/controld/controld_te_events.c

```diff
@@ -407,7 +447,13 @@ process_graph_event(xmlNode *event, const char *event_node)
 
     if (transition_num == -1) {
         // E.g. crm_resource --fail
+        if (g_hash_table_contains(outside_events, (gpointer) &action_num)) {
+            crm_debug("Outside event with transition key '%s' has already been "
+                      "processed", magic);
+            goto bail;
+        }
         desc = "initiated outside of the cluster";
+        g_hash_table_add(outside_events, (gpointer) &action_num);
```

Comment:
It's worth functionizing this to completely isolate the table handling. You can return something like EALREADY to indicate already processed. Since this will rarely be needed, it'll be better to create the table when needed in the new function. That also means null-checking the other uses.